ddev/src/ddev/e2e/agent/docker.py

```diff
@@ -0,0 +1,305 @@
+        if 'DD_API_KEY' not in env_vars:
+            # This fake key must be the proper length
+            env_vars['DD_API_KEY'] = 'a' * 32
+
+        # Set Agent hostname for CI
+        env_vars['DD_HOSTNAME'] = _get_hostname()
+
+        # Run API on a random free port
+        env_vars['DD_CMD_PORT'] = str(_find_free_port())
+
+        # Disable trace Agent
+        env_vars['DD_APM_ENABLED'] = 'false'
+
+        # Set up telemetry
+        env_vars['DD_TELEMETRY_ENABLED'] = '1'
+        env_vars['DD_EXPVAR_PORT'] = '5000'
+
+        # TODO: Remove this when Python 2 support is removed
+        #
+        # Don't write .pyc, needed to fix this issue (only Python 2):
+        # More info: https://github.com/DataDog/integrations-core/pull/5454
+        # When reinstalling a package, .pyc are not cleaned correctly. The issue is fixed by not writing them
+        # in the first place.
+        env_vars['PYTHONDONTWRITEBYTECODE'] = '1'
+
+        if (proxy_data := self.metadata.get('proxy')) is not None:
+            if (http_proxy := proxy_data.get('http')) is not None:
+                env_vars['DD_PROXY_HTTP'] = http_proxy
+            if (https_proxy := proxy_data.get('https')) is not None:
+                env_vars['DD_PROXY_HTTPS'] = https_proxy
+
+        volumes = []
+
+        if not self._is_windows_container:
+            volumes.append('/proc:/host/proc')
+
+        # Only mount the volume if the initial configuration is not set to `None`.
+        # As an example, the way SNMP does autodiscovery is that the Agent writes what its listener detects
+        # in `auto_conf.yaml`. The issue is we mount the entire config directory so changes are always in
+        # sync and, unlike other integrations that support autodiscovery, that file doesn't already exist.
+        # For that setup it seems the Agent cannot write to a file that does not already exist inside a
+        # directory that is mounted.
+        if self.config_file.is_file():
+            volumes.append(f'{self.config_file.parent}:{self._config_mount_dir}')
+
+        # It is safe to assume that the directory name is unique across all repos
+        for local_package in local_packages:
+            volumes.append(f'{local_package}:{self._package_mount_dir}{local_package.name}')
+
+        volumes.extend(self.metadata.get('docker_volumes', []))
+
+        if self.platform.windows and not self._is_windows_container:
+            for i, volume in enumerate(volumes):
+                parts = volume.split(':')
+                possible_file = ':'.join(parts[:2])
+                if os.path.isfile(possible_file):
+                    # Workaround for https://github.com/moby/moby/issues/30555
+                    vm_file = possible_file.replace(':', '', 1).replace('\\', '/')
+                    remaining = ':'.join(parts[2:])
+                    volumes[i] = f'/{vm_file}:{remaining}'
+
+        process = self._run_command(['docker', 'pull', agent_build])
```

Comment:
It may not be related to this PR directly but what if I want to use a docker image built locally? Would the `pull` command fail and the command just stops? (asking because it happened to me a few times and I just ended up commenting [this line](https://github.com/DataDog/integrations-core/blob/37f3a8b7c72d992f018c0f745d86932f75e8c7ba/datadog_checks_dev/datadog_checks/dev/tooling/commands/env/start.py#L80))

Reply:
I added an environment variable `DDEV_E2E_DOCKER_NO_PULL`